File: everything_parsing/28-grammar-tool-project2/js/firstsets.mjs
// Module 21 — Nullable and FIRST as LEAST FIXED POINTS (mirror of java/FirstSets.java).
//
// NULLABLE = smallest set N with:  A in N if A -> epsilon;  A in N if
// A -> X1..Xk and every Xi in N.  Start empty, apply until a pass adds nothing.
// FIRST is the same shape over sets of terminals. Both record every round.

export const EPS = "epsilon";

export class FirstSets {
  constructor(g) {
    this.g = g;
    this.nullableRounds = [];
    this.firstRounds = [];
    this.nullable = this._computeNullable();
    this.first = this._computeFirst();
  }

  _computeNullable() {
    const n = new Set();
    this.nullableRounds.push(new Set(n));
    let changed = true;
    while (changed) {
      changed = false;
      for (const p of this.g.productions) {
        if (n.has(p.lhs)) continue;
        let allNull = true;
        for (const s of p.rhs) { if (!n.has(s)) { allNull = false; break; } }
        if (allNull) { n.add(p.lhs); changed = true; }
      }
      this.nullableRounds.push(new Set(n));
    }
    return n;
  }

  nullableSeq(seq) {
    for (const s of seq) if (!this.nullable.has(s)) return false;
    return true;
  }

  _computeFirst() {
    const f = new Map();
    for (const nt of this.g.nonterminals) f.set(nt, new Set());
    this.firstRounds.push(snapshot(f));

    let changed = true;
    while (changed) {
      changed = false;
      for (const p of this.g.productions) {
        const target = f.get(p.lhs);
        const before = target.size;
        for (const x of this._firstOfSeqWith(f, p.rhs)) target.add(x);
        if (target.size !== before) changed = true;
      }
      this.firstRounds.push(snapshot(f));
    }
    return f;
  }

  _firstOfSeqWith(f, seq) {
    const out = new Set();
    let allNullable = true;
    for (const sym of seq) {
      if (!this.g.isNonterminal(sym)) { out.add(sym); allNullable = false; break; }
      for (const x of f.get(sym)) if (x !== EPS) out.add(x);
      if (!this.nullable.has(sym)) { allNullable = false; break; }
    }
    if (allNullable) out.add(EPS);
    return out;
  }

  firstOfSeq(seq) { return this._firstOfSeqWith(this.first, seq); }

  firstOf(sym) {
    if (!this.g.isNonterminal(sym)) return new Set([sym]);
    return new Set(this.first.get(sym));
  }

  firstIsStable() {
    const f = snapshot(this.first);
    for (const p of this.g.productions)
      for (const x of this._firstOfSeqWith(f, p.rhs)) f.get(p.lhs).add(x);
    for (const nt of this.g.nonterminals)
      if (!eqSet(f.get(nt), this.first.get(nt))) return false;
    return true;
  }
}

function snapshot(f) {
  const s = new Map();
  for (const [k, v] of f) s.set(k, new Set(v));
  return s;
}

function eqSet(a, b) {
  if (a.size !== b.size) return false;
  for (const x of a) if (!b.has(x)) return false;
  return true;
}

export function setStr(s) {
  const items = [...s].sort((a, b) => {
    const ae = a === EPS, be = b === EPS;
    if (ae !== be) return ae ? 1 : -1;
    return a < b ? -1 : a > b ? 1 : 0;
  });
  return "{ " + items.join(" ") + " }";
}

File: everything_parsing/22-follow-sets/java/FirstSets.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * Module 21 — Nullable and FIRST as LEAST FIXED POINTS.
 *
 * NULLABLE is the smallest set of nonterminals N such that:
 *     A in N   if   A -> epsilon
 *     A in N   if   A -> X1 X2 ... Xk  and every Xi is in N
 * You cannot just read this off — the second rule refers to N on both sides.
 * Start from the empty set (nothing assumed nullable) and apply the rules until
 * a full pass adds nothing. Each pass can only ADD (the rules are monotone) and
 * there are finitely many nonterminals, so it stops. The result is a fixed
 * point, and because we grew from below it is the LEAST one.
 *
 * FIRST is the same shape, over sets of terminals:
 *     FIRST(A) contains  a               if  A -> a ...           (a terminal)
 *     FIRST(A) contains  FIRST(Xi)       if  A -> X1..Xi.. and X1..X(i-1) nullable
 *     FIRST(A) contains  epsilon         if  A -> X1..Xk and all Xi nullable
 *
 * Both computations record every round so `Main` can print the trace.
 */
public final class FirstSets {

    public static final String EPS = "epsilon";

    public final Grammar g;
    public final Set<String> nullable;
    public final List<Set<String>> nullableRounds = new ArrayList<>();
    public final Map<String, Set<String>> first;
    public final List<Map<String, Set<String>>> firstRounds = new ArrayList<>();

    public FirstSets(Grammar g) {
        this.g = g;
        this.nullable = computeNullable();
        this.first = computeFirst();
    }

    // ─────────────────────────────────────────── NULLABLE

    private Set<String> computeNullable() {
        Set<String> n = new LinkedHashSet<>();
        nullableRounds.add(new LinkedHashSet<>(n));         // round 0: {}
        boolean changed = true;
        while (changed) {
            changed = false;
            for (Grammar.Production p : g.productions) {
                if (n.contains(p.lhs())) continue;
                boolean allNull = true;
                for (String s : p.rhs()) {
                    if (!n.contains(s)) { allNull = false; break; }   // a terminal, or a not-yet-nullable NT
                }
                if (allNull) { n.add(p.lhs()); changed = true; }      // empty rhs => allNull stays true
            }
            nullableRounds.add(new LinkedHashSet<>(n));
        }
        return n;
    }

    public boolean nullableSeq(List<String> seq) {
        for (String s : seq) if (!nullable.contains(s)) return false;
        return true;
    }

    // ─────────────────────────────────────────── FIRST

    private Map<String, Set<String>> computeFirst() {
        Map<String, Set<String>> f = new LinkedHashMap<>();
        for (String nt : g.nonterminals) f.put(nt, new LinkedHashSet<>());
        firstRounds.add(snapshot(f));                       // round 0: all empty

        boolean changed = true;
        while (changed) {
            changed = false;
            for (Grammar.Production p : g.productions) {
                Set<String> target = f.get(p.lhs());
                int before = target.size();
                target.addAll(firstOfSeqWith(f, p.rhs()));
                if (target.size() != before) changed = true;
            }
            firstRounds.add(snapshot(f));
        }
        return f;
    }

    private Set<String> firstOfSeqWith(Map<String, Set<String>> f, List<String> seq) {
        Set<String> out = new LinkedHashSet<>();
        boolean allNullable = true;
        for (String sym : seq) {
            if (!g.isNonterminal(sym)) { out.add(sym); allNullable = false; break; }
            for (String x : f.get(sym)) if (!x.equals(EPS)) out.add(x);
            if (!nullable.contains(sym)) { allNullable = false; break; }
        }
        if (allNullable) out.add(EPS);
        return out;
    }

    /** FIRST of a symbol string, using the finished tables. */
    public Set<String> firstOfSeq(List<String> seq) { return firstOfSeqWith(first, seq); }

    public Set<String> firstOf(String sym) {
        if (!g.isNonterminal(sym)) return new LinkedHashSet<>(List.of(sym));
        return new LinkedHashSet<>(first.get(sym));
    }

    // ─────────────────────────────────────────── "is this really a fixed point?"

    /** Run one more FIRST pass; return true iff nothing changes (i.e. we are at a fixed point). */
    public boolean firstIsStable() {
        Map<String, Set<String>> f = snapshot(first);
        for (Grammar.Production p : g.productions)
            f.get(p.lhs()).addAll(firstOfSeqWith(f, p.rhs()));
        for (String nt : g.nonterminals)
            if (!f.get(nt).equals(first.get(nt))) return false;
        return true;
    }

    // ─────────────────────────────────────────── rendering

    private static Map<String, Set<String>> snapshot(Map<String, Set<String>> f) {
        Map<String, Set<String>> s = new LinkedHashMap<>();
        for (var e : f.entrySet()) s.put(e.getKey(), new LinkedHashSet<>(e.getValue()));
        return s;
    }

    public static String setStr(Set<String> s) {
        TreeSet<String> t = new TreeSet<>((a, b) -> {
            boolean ae = a.equals(EPS), be = b.equals(EPS);
            if (ae != be) return ae ? 1 : -1;
            return a.compareTo(b);
        });
        t.addAll(s);
        return "{ " + String.join(" ", t) + " }";
    }
}
